src/Data/HashMap/Linear.hs

```diff
@@ -59,218 +58,185 @@ import qualified Prelude
+  HashMap :: Size -> Count -> RobinArr k v #-> HashMap k v
 
 -- INVARIANTS:
 --   * Cells are empty iff the PSL is -1.
---   * We are sparse: count*3 <= size, with correct values of count and size
 --   * Each (RobinVal k v) has the correct PSL
+--   * We NEVER evaluate the key-val pair for PSL -1
+
+-- | The size allocated for the array
+newtype Size = Size Int
+  deriving (Prelude.Num)
+
+-- | The number of pairs stored in the hashmap
+newtype Count = Count Int
+  deriving (Prelude.Num)
+
+-- | An array of Robin values
+type RobinArr k v = Array (RobinVal k v)
+
+-- | Robin values are triples of the key, value and PSL
+-- (the probe sequence length).
+type RobinVal k v = (PSL, (k, v))
+
+-- | A probe sequence length
+newtype PSL = PSL Int
+  deriving (Prelude.Eq, Prelude.Ord, Prelude.Num, Prelude.Show)
+
+-- | At minimum, we need to store hashable
+-- and identifiable keys
+type Keyed k = (Eq k, Hashable k)
 
 -- | The results of searching for where to insert a key
-data RobinQuery k where
+data ProbeResult k where
   -- | A key, PSL pair to insert at an index, that has not
   -- yet been touched. Invariant: the PSL must be -1 at that index.
-  IndexToInsert :: (k, PSL) -> Int -> RobinQuery k
+  IndexToInsert :: (k, PSL) -> Int -> ProbeResult k
   -- A key, PSL pair to update the value at an index.
-  IndexToUpdate :: (k, PSL) -> Int -> RobinQuery k
+  IndexToUpdate :: (k, PSL) -> Int -> ProbeResult k
   -- | A key, PSL pair for a swap at an index.
   -- The swapped-out pair will then need to be inserted downstream.
-  IndexToSwap :: (k, PSL) -> Int -> RobinQuery k
+  IndexToSwap :: (k, PSL) -> Int -> ProbeResult k
+
 
 -- # Construction and Modification
 --------------------------------------------------
 
--- | Run a computation using a singleton hashmap
-singleton :: Keyed k =>
-  (k, v) -> (HashMap k v #-> Unrestricted b) -> Unrestricted b
-singleton (k :: k, v :: v) (f :: HashMap k v #-> Unrestricted b) =
-  alloc defaultSize (k, v, -1) applyHM
-  where
-    applyHM :: Array (RobinVal k v) #-> Unrestricted b
-    applyHM arr = let ixToHash = (hash k) `mod` defaultSize in
-      f $ HashMap (defaultSize, 1) (write arr ixToHash (k, v, 0))
-
-singleton' :: Keyed k =>
-  (k, v) -> (HashMap k v #-> Unrestricted b) -> b
-singleton' kv f = unUnrestricted (singleton kv f)
-
--- XXX: re-write linearly
--- | Given a key @k@ to 'lookup', look up a @Maybe v@ and feed it to
--- a function @Maybe v -> Maybe v@, and if the result is @Nothing@,
--- delete that key, and otherwise, replace the value with the @v@
--- in the returned @Just v@.
-alter ::
-  Keyed k =>
-  HashMap k v #-> (Maybe v -> Maybe v) ->
-  k ->
-  HashMap k v
-alter = Unsafe.toLinear unsafeAlter
+-- | Run a computation given an empty hashmap
+empty :: forall k v b.
+  Keyed k => Size -> (HashMap k v #-> Unrestricted b) -> Unrestricted b
+empty sz@(Size s) scopeFn =
+  alloc s (-1, (error "key err val", error "val err val")) robinArrayScopeFn
```

Comment:
There error messages are really cryptic. Just put a real sentence there. It becomes self-documenting. Better for the two error messages to be different indeed, as they will be easier to tell apart.